dev/src/write-batch.ts

```diff
@@ -589,15 +575,198 @@ export class WriteBatch {
+            arg, 'precondition')} Input contains more than one condition.`);
+  }
+}
+
+
+/**
+ * Validates the use of 'value' as an update Precondition.
+ *
+ * @private
+ * @param arg The argument name or argument index (for varargs methods).
+ * @param value The object to validate.
+ * @param options Optional validation options specifying whether the valued can
+ * be omitted.
+ */
+function validateUpdatePrecondition(
+    arg: string|number, value: unknown, options?: AllowOptional): void {
+  if (!validateOptional(value, options)) {
+    validatePrecondition(arg, value, /* allowExists= */ false);
+  }
+}
+
+/**
+ * Validates the use of 'value' as a delete Precondition.
+ *
+ * @private
+ * @param arg The argument name or argument index (for varargs methods).
+ * @param value The object to validate.
+ * @param options Optional validation options specifying whether the valued can
+ * be omitted.
+ */
+function validateDeletePrecondition(
+    arg: string|number, value: unknown, options?: AllowOptional): void {
+  if (!validateOptional(value, options)) {
+    validatePrecondition(arg, value, /* allowExists= */ true);
+  }
+}
+
+/**
+ * Validates the use of 'value' as SetOptions and enforces that 'merge' is a
+ * boolean.
+ *
+ * @private
+ * @param arg The argument name or argument index (for varargs methods).
+ * @param value The object to validate.
+ * @param options Optional validation options specifying whether the valued can
+ * be omitted.
+ * @returns 'true' if the input is a valid SetOptions object.
+ */
+export function validateSetOptions(
+    arg: string|number, value: unknown, options?: AllowOptional): void {
+  if (!validateOptional(value, options)) {
+    if (!isObject(value)) {
+      throw new Error(`${
+          invalidArgumentMessage(
+              arg, 'set() option')} Input is not an object.`);
+    }
+
+    const setOptions = value as {[k: string]: unknown};
+
+    if ('merge' in setOptions && typeof setOptions.merge !== 'boolean') {
+      throw new Error(`${
+          invalidArgumentMessage(
+              arg, 'set() option')} "merge" is not a boolean.`);
+    }
+
+    if ('mergeFields' in setOptions) {
+      if (!Array.isArray(setOptions.mergeFields)) {
+        throw new Error(`${
+            invalidArgumentMessage(
+                arg, 'set() option')} "mergeFields" is not an array.`);
+      }
+
+      for (let i = 0; i < setOptions.mergeFields.length; ++i) {
+        try {
+          validateFieldPath(i, setOptions.mergeFields[i]);
+        } catch (err) {
+          throw new Error(`${
+              invalidArgumentMessage(
+                  arg,
+                  'set() option')} "mergeFields" is not valid: ${err.message}`);
+        }
+      }
+    }
+
+    if ('merge' in setOptions && 'mergeFields' in setOptions) {
+      throw new Error(`${
+          invalidArgumentMessage(
+              arg,
+              'set() option')} You cannot specify both "merge" and "mergeFields".`);
+    }
+  }
```

Comment:
Again, should we be checking for extra invalid items in setOptions?  FWIW- the client SDK has `validateOptionNames()` to do that.  You could also consider borrowing `validateNamedOptionalType()` to replace your manual type-checking of 'merge', 'mergeFields' etc.  Up to you though.